### PR TITLE
TW-1955: Create permission dialogs for contacts and images

### PR DESCRIPTION
**Pull request by Te-Z**
## Ticket

#1955 

Evolution of Google Privacy policies need dialog before asking contact and image permission

## Resolved
- Android:

https://github.com/user-attachments/assets/12e2caa7-3def-4bf3-8bb1-bd1f8fde6b7d

https://github.com/user-attachments/assets/6d41d3a4-bbcb-4eb6-aba6-496fdefb682b

- IOS:

https://github.com/user-attachments/assets/0b40c790-4b6b-42e1-8264-5b33a7df0c39

https://github.com/user-attachments/assets/5db61025-517a-41c9-bbac-19c16c9e663d

https://github.com/user-attachments/assets/ecfe01d0-8f9c-4219-9254-1e032b1a6b35

### CODE REVIEW

**Comment by nqhhdev** [comment]
IMO: 
- Currently we have a warning banner for contact permission in contact tab. You should re-system the banner and dialog consistently
- When we request contact permission:
   - if permission is not `Granted` → show dialog
   - if permission is `Deny` → don't need to show dialog, only show warning banner
   
 For example:

 - Install app Twake Chat and logged.
 - Open contact tabs → permission is not `Granted` → show dialog then click `Deny` on Twake Dialog to close.
 - Open again contact tabs → permission is `Deny` → Not show dialog, only show warning banner → continue actions
 
 Simulator Screenshot - iPhone 15 - 2024-07-18 at 11 35 05

**Comment by nqhhdev** [comment]
One more thing: 
- When we display permission dialog first → click on actions (Deny or Allow) → close dialog and request native permission → show native dialog → continue

File: lib/pages/contacts_tab/contacts_tab.dart [hunk 1]
```diff
@@ -45,6 +58,17 @@ class ContactsTabController extends State<ContactsTab>
 
     _listenFocusTextEditing();
     super.initState();
+    WidgetsBinding.instance.addObserver(this);
+  }
+
+  @override
+  void didChangeAppLifecycleState(AppLifecycleState state) {
+    if (state == AppLifecycleState.resumed) {
+      initialFetchContacts(
```
Comment (nqhhdev) [inline lib/pages/contacts_tab/contacts_tab.dart:67]: Why we need to fetch contacts?

Comment (nqhhdev) [reply lib/pages/contacts_tab/contacts_tab.dart:67]: Side effect

https://github.com/user-attachments/assets/09dab8b3-ebd5-4d76-9dba-e5ac4b9b4d1d

Comment (Te-Z) [reply lib/pages/contacts_tab/contacts_tab.dart]: Not a side effect.

File: lib/pages/contacts_tab/contacts_tab.dart [hunk 2]
```diff
@@ -36,6 +39,13 @@ class ContactsTabController extends State<ContactsTab>
   void initState() {
     SchedulerBinding.instance.addPostFrameCallback((_) async {
       if (mounted) {
+        if (widget.autoDisplayPermissionContactDialog.value) {
+          await displayPermissionDialog(
+            context: context,
+            autoDisplayNotifier: widget.autoDisplayPermissionContactDialog,
+          );
+        }
+
```
Comment (nqhhdev) [inline lib/pages/contacts_tab/contacts_tab.dart:48]: Should handle it's on `ContactsViewControllerMixin`

File: lib/widgets/layouts/adaptive_layout/app_adaptive_scaffold_body.dart [hunk 3]
```diff
@@ -43,6 +43,9 @@ class AppAdaptiveScaffoldBodyController extends State<AppAdaptiveScaffoldBody>
 
   final activeRoomIdNotifier = ValueNotifier<String?>(null);
 
+  final ValueNotifier<bool> autoDisplayPermissionContactDialog =
```
Comment (nqhhdev) [inline lib/widgets/layouts/adaptive_layout/app_adaptive_scaffold_body.dart:46]: -  I don't see this variable update, always is true. 
- Missing dispose 

IMO: No need variable.

File: lib/utils/permission_service.dart [hunk 4]
```diff
@@ -136,7 +217,7 @@ class PermissionHandlerService {
     return await Permission.photosAddOnly.request();
   }
 
-  void goToSettingsForPermissionActions() {
-    openAppSettings();
+  Future goToSettingsForPermissionActions() async {
```
Comment (nqhhdev) [inline lib/utils/permission_service.dart:220]: IMO: No need to change `Future` func

File: lib/presentation/mixins/contacts_view_controller_mixin.dart [hunk 5]
```diff
@@ -40,7 +43,7 @@ mixin class ContactsViewControllerMixin {
       getIt.get<SearchRecentChatInteractor>();
 
   ValueNotifier<WarningContactsBannerState> warningBannerNotifier =
-      ValueNotifier(WarningContactsBannerState.hide);
+      ValueNotifier(WarningContactsBannerState.display);
```
Comment (nqhhdev) [inline lib/presentation/mixins/contacts_view_controller_mixin.dart:46]: - Why do you change variable? 
- Side effect from this:

https://github.com/user-attachments/assets/0c3c78c0-0a2a-4dd6-ad78-2f0694852545

File: lib/presentation/mixins/contacts_view_controller_mixin.dart [hunk 6]
```diff
@@ -67,16 +70,41 @@ mixin class ContactsViewControllerMixin {
...
+            permission: Permission.contacts,
+            explainTextRequestPermission: Text(
+              L10n.of(context)!.explainPermissionToAccessContacts,
+            ),
+            onAcceptButton: () async {
+              dialogContext.pop();
+
+              await _permissionHandlerService
+                  .requestContactsPermissionActions();
+              await handleRequestContactsPermission();
```
Comment (nqhhdev) [inline lib/presentation/mixins/contacts_view_controller_mixin.dart:95]: IMO: Should create a func to handle when click on button

File: lib/utils/permission_service.dart [hunk 7]
```diff
@@ -55,28 +61,78 @@ class PermissionHandlerService {
...
+            onAcceptButton: () async {
+              try {
+                await Permission.photos.request();
+              } finally {
+                dialogContext.pop();
+              }
+            },
+          );
+        },
+      );
```
Comment (nqhhdev) [inline lib/utils/permission_service.dart:107]: IMO: Duplicated code, pls separate a func can be reuse

File: lib/presentation/mixins/contacts_view_controller_mixin.dart [hunk 8]
```diff
@@ -67,16 +70,41 @@ mixin class ContactsViewControllerMixin {
...
+        context: context,
+        builder: (dialogContext) {
+          return PermissionDialog(
+            icon: const Icon(Icons.contact_page_outlined),
+            permission: Permission.contacts,
+            explainTextRequestPermission: Text(
+              L10n.of(context)!.explainPermissionToAccessContacts,
+            ),
+            onAcceptButton: () async {
+              dialogContext.pop();
```
Comment (nqhhdev) [inline lib/presentation/mixins/contacts_view_controller_mixin.dart:91]: SHOULD use Navigator to pop previous screen. NOT use `go_router`

File: lib/utils/permission_dialog.dart [hunk 9]
```diff
@@ -74,6 +78,7 @@ class _PermissionDialogState extends State<PermissionDialog>
                       context: context,
                       text: L10n.of(context)!.deny,
                       onPressed: () {
+                        widget.onRefuseTap!();
```
Comment (nqhhdev) [inline lib/utils/permission_dialog.dart:81]: BAD GOOD ! If `onRefuseTap` is null → CRASH APP

File: lib/utils/permission_service.dart [hunk 10]
```diff
@@ -55,28 +61,78 @@ class PermissionHandlerService {
...
+            permission: Permission.contacts,
+            explainTextRequestPermission: Text(
+              L10n.of(context)!.explainPermissionToAccessPhotos,
+            ),
+            onAcceptButton: () async {
+              try {
+                await Permission.photos.request();
+              } finally {
+                dialogContext.pop();
+              }
```
Comment (nqhhdev) [inline lib/utils/permission_service.dart:103]: Should close `PermissionDialog` before requesting native permission

File: lib/pages/new_group/contacts_selection.dart [hunk 11]
```diff
@@ -34,22 +37,34 @@ abstract class ContactsSelectionController<T extends StatefulWidget>
...
+
   @override
   void initState() {
     SchedulerBinding.instance.addPostFrameCallback((_) async {
+      WidgetsBinding.instance.addObserver(this);
       if (mounted) {
         initialFetchContacts(
-          client: client,
+          context: context,
+          client: Matrix.of(context).client,
```
Comment (hoangdat) [inline lib/pages/new_group/contacts_selection.dart:49]:
```suggestion
          client: client,
```

File: lib/pages/new_group/contacts_selection.dart [hunk 12]
```diff
@@ -34,22 +37,34 @@ abstract class ContactsSelectionController<T extends StatefulWidget>
 
   Client get client => Matrix.of(context).client;
 
+  final autoDisplayPermissionDialogNotifier = ValueNotifier(true);
```
Comment (hoangdat) [inline lib/pages/new_group/contacts_selection.dart:40]: remove it

File: lib/pages/new_private_chat/widget/expansion_list.dart [hunk 13]
```diff
@@ -164,7 +163,6 @@ class ExpansionList extends StatelessWidget {
                     const SizedBox(
                       height: 12,
                     ),
-                    _contactsWarningBannerViewBuilder(),
```
Comment (hoangdat) [inline lib/pages/new_private_chat/widget/expansion_list.dart:167]: why need to remove it?

Comment (hoangdat) [reply lib/pages/new_private_chat/widget/expansion_list.dart:167]: @Te-Z

Comment (nqhhdev) [reply lib/pages/new_private_chat/widget/expansion_list.dart:167]: Reverted

**Review by nqhhdev — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**